fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggThroughJoinOnPkFk.java

```diff
@@ -0,0 +1,317 @@
+        HashMap<Slot, Slot> primaryToForeignDeps = new HashMap<>();
+        for (Slot slot : primarySlots) {
+            Set<Set<Slot>> replacedSlotSets = funcDeps.findDeterminats(ImmutableSet.of(slot));
+            for (Set<Slot> replacedSlots : replacedSlotSets) {
+                if (primaryOutputSet.stream().noneMatch(replacedSlots::contains)
+                        && replacedSlots.size() == 1) {
+                    primaryToForeignDeps.put(slot, replacedSlots.iterator().next());
+                    break;
+                }
+            }
+        }
+
+        Set<Expression> newGroupBySlots = constructNewGroupBy(agg, primaryOutputSet, primaryToForeignDeps);
+        List<NamedExpression> newOutput = constructNewOutput(agg, primaryOutputSet, primaryToForeignDeps, funcDeps);
+        if (newGroupBySlots == null || newOutput == null) {
+            return null;
+        }
+        return agg.withGroupByAndOutput(ImmutableList.copyOf(newGroupBySlots), ImmutableList.copyOf(newOutput));
+    }
+
+    private @Nullable Set<Expression> constructNewGroupBy(LogicalAggregate<?> agg, Set<Slot> primaryOutputs,
+            Map<Slot, Slot> primaryToForeignBiDeps) {
+        Set<Expression> newGroupBySlots = new HashSet<>();
+        for (Expression expression : agg.getGroupByExpressions()) {
+            if (!(expression instanceof Slot)) {
+                return null;
+            }
+            if (primaryOutputs.contains((Slot) expression)
+                    && !primaryToForeignBiDeps.containsKey((Slot) expression)) {
+                return null;
+            }
+            expression = primaryToForeignBiDeps.getOrDefault(expression, (Slot) expression);
+            newGroupBySlots.add(expression);
+        }
+        return newGroupBySlots;
+    }
+
+    private @Nullable List<NamedExpression> constructNewOutput(LogicalAggregate<?> agg, Set<Slot> primaryOutput,
+            Map<Slot, Slot> primaryToForeignDeps, FuncDeps funcDeps) {
+        List<NamedExpression> newOutput = new ArrayList<>();
+        for (NamedExpression expression : agg.getOutputExpressions()) {
+            // There are three cases for output expressions:
+            // 1. Slot: the slot is from primary plan, we need to replace it with
+            //             the corresponding slot from foreign plan,
+            //             or skip it when it isn't in group by.
+            // 2. Count: the count is from primary plan,
+            //             we need to replace the slot in the count with the corresponding slot
+            //             from foreign plan
+            // 3. Others: the expression is not from primary plan, we need to keep it
+            if (expression instanceof Slot && primaryToForeignDeps.containsKey(expression)) {
+                expression = primaryToForeignDeps.getOrDefault(expression, expression.toSlot());
+            }
+            if (expression instanceof Alias
+                    && expression.child(0) instanceof Count
```

Comment:
why only count?

Reply:
Because count only cares about row count. The circle dependenciBecause COUNT only cares about row count, the circular dependencies mean that if two rows in column 1 are the same, then they are also the same in column 2. So their row counts are the same, but their values can be different